### ISSUE TITLE
Remove overlap between hpc-shell and "The Unix Shell"

### ISSUE THREAD

**Issue by reid-a** [hpc-carpentry/hpc-shell#18]
In light of some of the previous issues, this might be a bit fraught, but it seems that this lesson has a lot of overlap with the HPC Carpentry "core" [shell lesson](http://swcarpentry.github.io/shell-novice/). That lesson is maintained by a large team of motivated volunteers, which makes it both high quality and a moving target. Components of the hpc-shell lesson taken from that lesson likely age quickly, and suffer in comparison, and remedying this circumstance would require a high level of duplicated effort.

However, the HPC environment is very shell-centric, and in particular makes more sophisticated use of some of the tools (ssh and shell scripts particularly) than is covered in the shell lesson, particularly for those learners whose shell lesson exposure might have been cut short due to time constraints.

So the proposal is not to discard this lesson, but to pare it down to consist exclusively of shell-specific items which are not covered in the Carpentry shell lesson.

On a first comparison (undertaken a few days ago by @tkphd and I), it appears the sections already present in this lesson which would be retained are:

- Why use a cluster?
- Opening a terminal on your local system, because it includes puTTY, implied in Windows SSH.
- Connecting via SSH
- Local vs remote systems
- HPC file systems -- differences in sharing, protection, performance.
- File system permissions, read vs execute, chmod (covered lightly in the non-hpc shell)
- Editors -- you might encounter vim in addition to nano
- File extensions are arbitrary
- Files can be moved between machines -- wget
- Files can be large -- compression, decompression, alternatives to 'cat'
- Default devices -- stdin, stdout, stderr, use of `&>` for redirection 
- Scripts can be in files, meaning of `#!/bin/bash`
- Paths and permissions
- Use of `$(cmd)` to populate shell environment variables with results of a command

With this abbreviated hpc-shell lesson, the plan to get learners from zero to running on the HPC system would be, first, teach the core Carpentries Unix shell intro, then use this new abbreviated hpc-shell as a supplement, and then go on with hpc-intro.

This seems like a way of capturing all the important content for our learners, while minimizing duplicated effort.

Whether to merge the resulting abbreviated hpc-shell lesson with hpc-intro remains an open question, I don't think I have a strong opinion about that.

**Comment by CarrieBrown** [comment]
Hi Andy,

This seems like a great idea to keep in mind moving forward. I completely understand and support the desire to not maintain redundant information. However, as evidenced by the Data Carpentry lessons, there are benefits to maintaining alternate shell lessons. The benefits of offering an HPC-specific shell lesson are worth the possible pain of maintaining the lesson.

**Teaching the lesson "on the hardware" streamlines setup for the learners.** Having to install both local shell software in addition to HPC-specific software (especially for Windows users) is a lot to ask of a novice user. Initial setup can be a challenging thing and doubling that load is a lot to ask.

**Keeping the lesson HPC-specific also allows us to focus more on these topics that are more important on HPC systems without adding additional learning time.** One can cover things like file permissions and chmod in HPC specific examples in place of some of the existing examples in the shell lesson. Some of the topics specified are best contained as a callout (such as the use of vim or `&>` for redirection) so they can be included or skipped based on how advanced the group of learners you are currently teaching.

Lastly, we need to look at **the shortest path to the goal, which is to get these lessons integrated into The Carpentries.** A large-scale overhaul at this time would just lengthen this process.

If we do hope to have HPC Carpentry be its own unit of The Carpentries (something that I think would be a great goal as there are ALOT of HPC-centric topics that are just begging to be put into a Carpentries style lesson, such as Parallelization and Containerization) it would be to our benefit to have a standalone `hpc-shell` lesson.

**Comment by aturner-epcc** [comment]
I have had a look through both lessons and agree with @reid-a and @tkphd as to the correct approach here. I also agree with @CarrieBrown that we can defer until HPC Carpentry is incorporated into The Carpentries with the caveat that I do not yet have an understanding of how long incorporating HPC Carpentry will take. I think, roughly, If it is going to take under 6 months to incorporate HPC Carpentry into the Carpentries then we can defer this great idea until this is complete. If the incorporation process is going to take longer than this then I do not think we can stall on making progress for that long given the momentum we currently have.

Looking at the overlap between the shell and hpc-shell lessons I cannot see a strong argument for keeping a separate hpc-shell. In fact, I think that pulling out the HPC-specific topics into a shorter lesson will be useful in giving us more flexibility for structuring workshops and fits well with the modular approach that we are taking for HPC Carpentry. For example, this separation would allow us to meet more nuanced novice HPC learner needs based on their current experience:

* Never used bash shell: shell + hpc-shell + hpc-intro
* Used bash shell locally: hpc-shell + hpc-intro
* Used remote systems via shell (e.g. cloud resources): hpc-intro

I definitely like this structure and I think it maps well onto the correct model for partitioning the required skills.

**Comment by DrDaveTurner** [comment]
I agree fully that having multiple Unix introductions is a problem.  However, I disagree on the proposal that we pare down the HPC Unix intro and make it a supplement to Unix Novice.  This is mainly due to the fact that shell-novice has many deficiencies at this point, and I don't think that HPC Unix intro has nearly as many.  In Unix novice, most examples are given for OS-X, which is not what we want in the HPC community, and OS-X is moving from Bash to Zsh going forward anyway.  Flags are introduced using ls -F which is not needed in Linux since ls automatically colorizes the output.

I think we'd be better off in the short run with fully developing HPC Unix Intro ourselves.  I think we should also keep in mind an eye toward generalizing it so that it could be offered as a potential replacement for Unix novice, where advance or HPC specific material could be pared back and more exercises used for instructors who want a more generalized Unix introduction.  Some of my reasoning for this is that I think any Unix introduction needs to include a full discussion of running shells with #!/bin/bash and permissions which Unix Novice ignores.  Another reason this makes sense to me is that the existing teaching environment is changing rapidly, with OS-X moving from Bash to Zsh and with Windows having native support for Linux.  Looking forward this means we mostly need to support Linux as the mainstream with Mac's being the outlier for a general Unix course.

So basically I think the the HPC Unix Introduction material is a better starting place at this point.  I do realize that it would be very difficult to convince the group who has invested a great deal of effort into developing shell-novice into adopting HPC Unix

**Comment by bkmgit** [comment]
Possibly material on logging in can be moved/replicated in hpc-intro, though some people may not use Software Carpentry shell lesson. A related issue is https://github.com/carpentries-incubator/hpc-intro/issues/333
